### DISCUSSION TITLE
Writing cache for long running computation

### DISCUSSION THREAD

**Discussion by dolzenko** [ruby-concurrency/concurrent-ruby#1010] [Q&A] [answered]
Hi, first of all thanks for the great gem! Now I know this is hinted at in the docs but I was looking for examples, the problem is - I'm writing a cache for long running computation based on Concurrent::Map and I want to make sure only single thread runs the computation in case the value is absent (with other threads waiting for result from this single thread), is it possible using just the Concurrent::Map (as far as I can tell it isn't), and if not what primitives are the best/most robust to choose?

**Comment by bensheldon** [comment]
`Concurrent::Map#compute_if_absent`'s [documentation](https://ruby-concurrency.github.io/concurrent-ruby/master/Concurrent/Map.html#compute_if_absent-instance_method) describes it as "atomic", meaning it should only execute on one thread. If you're running CRuby, you can see that the computation part is synchronized:

https://github.com/ruby-concurrency/concurrent-ruby/blob/9f40827be9a8a192a6993a8d157bd3ed0662ada0/lib/concurrent-ruby/concurrent/collection/map/mri_map_backend.rb#L21C1-L27

**Comment by dolzenko** [comment]
Ok, seems like I misunderstood the docs somewhat, is there any place I can read up on the definition of "atomic"?

**Reply by bensheldon** [reply, marked as answer]
It took me a long while to become familiar with the library 🤗 

Concurrent Ruby borrows quite a lot from Java's concurrency primitives. You might start there: https://docs.oracle.com/javase/tutorial/essential/concurrency/atomic.html

**Reply by dolzenko** [reply]
Ok, this works for me then for now, now gotta figure out why all these concurrent calls in my code...
